allennlp/data/vocabulary.py

```diff
@@ -285,6 +287,40 @@ def from_dataset(cls,
                           max_vocab_size=max_vocab_size,
                           non_padded_namespaces=non_padded_namespaces)
 
+    @classmethod
+    def from_params(cls, params: Params, dataset=None):
+        """
+        There are two possible ways to build a vocabulary; from a
+        pre-existing dataset, using :func:`Vocabulary.from_dataset`, or
+        from a pre-saved vocabulary, using :func:`Vocabulary.from_files`.
+        This method wraps both of these options, allowing their specification
+        from a ``Params`` object, generated from a JSON configuration file.
+
+        Parameters
+        ----------
+        params: Params, required.
+        dataset: Dataset, optional.
+            If ``params`` doesn't contain a ``vocabulary_directory`` key,
+            the ``Vocabulary`` can be built directly from a ``Dataset``.
+
+        Returns
+        -------
+        A ``Vocabulary``.
+        """
+        vocabulary_directory = params.pop("vocabulary_directory", None)
+        if not vocabulary_directory and not dataset:
+            raise ConfigurationError("You must provide either a Params object containing a "
+                                     "vocab_directory key or a Dataset to build a vocabulary from.")
+        if vocabulary_directory and dataset:
+            logger.warning("from_params was passed both a vocabulary_directory and a Dataset. This "
+                           "behaviour is not well defined, so backing off to using the vocabulary "
+                           "defined at %s.", vocabulary_directory)
+
+        if vocabulary_directory:
+            params.assert_empty("Vocabulary - from files")
+            return Vocabulary.from_files(vocabulary_directory)
```

Comment:
I don't think there's a reason to create this `from_files` here - you can just call `Vocabulary.from_files` if that's what you want.

Reply:
I don't understand? This is required to be able to specify a pre-made vocab when using `allennlp/run.py train`, because otherwise we have to hard code in `Vocabulary.from_dataset` in `train`.

Reply:
Do we need to specify a pre-made vocab?

Reply:
I'd imagine this would be how we do inference with a different set of glove vectors - use a different pre-specified vocab, and then load pretrained embeddings given this new vocabulary. We can't do that at the moment because of the way we disregard pretrained embeddings when we load models, but if that was a little more sophisticated, I could easily see this being how it works. Also, having this would speed things up - write a vocabulary once for the first run and then specify it in all other runs.

Reply:
If the goal is to make this work with different glove vectors, I think it'd be smarter to wait on this change until we have a whole solution for doing that - it's likely that actually trying to get that use case to work will want a different solution than what you have here.

Also, I think we can come up with a better way to handle caching pre-processed inputs that doesn't require modifying parameter files.  Maybe...

...

Ok, after thinking about this some more, I think it's fine to do it this way.  It seems reasonable to want to do this.  The logic in here is a little more complicated than I'd like, but I don't have a better option.